src/main/java/com/trainigcenter/springtask/web/controller/GenreController.java

```diff
@@ -0,0 +1,130 @@
+    public List<GenreDto> getAll() {
+        List<Genre> allGenres = genreService.getAll();
+        return allGenres.stream()
+                        .map(this::convertToDto)
+                        .collect(Collectors.toList());
+    }
+
+    @GetMapping("/{id}")
+    public GenreDto getGenreById(@PathVariable("id") int id) {
+        Genre genre = Optional.ofNullable(genreService.getById(id))
+                              .orElseThrow(() -> new NotFoundException("Genre id:" + id + " not found"));
+
+        return convertToDto(genre);
+    }
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public GenreDto saveGenre(@Valid @RequestBody GenreDto genreDto) {
+        genreDto.setId(null);
+        genreDto.setName(mapGenreName(genreDto.getName()));
+
+        Genre genre = genreService.add(convertFromDto(genreDto));
+        return convertToDto(genre);
+    }
+
+    @PutMapping("/{id}")
+    @ResponseStatus(HttpStatus.CREATED)
+    public GenreDto updateGenre(@PathVariable("id") Integer id,
+                                @Valid @RequestBody GenreDto genreDto) {
+
+        Genre genre = convertFromDto(genreDto);
+        genre.setId(id);
+        genre.setName(mapGenreName(genre.getName()));
+
+        genre = Optional.ofNullable(genreService.update(genre))
+                        .orElseThrow(() -> new NotFoundException("Genre id:" + id + " not found"));
+
+        return convertToDto(genre);
+    }
+
+    @DeleteMapping("/{id}")
+    public void deleteGenre(@PathVariable("id") Integer id) {
+
+        Genre genre = Optional.ofNullable(genreService.getByIdWithMovies(id))
+                              .orElseThrow(() -> new NotFoundException("Genre id:" + id + " not found"));
+
+        boolean isDeleted = genreService.delete(genre);
+
+        if (!isDeleted) {
+            throw new ForbiddenException(genre.getName() + " is take part in movies");
+        }
+
+    }
+
+    @ExceptionHandler({NotFoundException.class, ForbiddenException.class, MethodNotAllowedException.class})
+    private ResponseEntity<Error> exceptionHandler(WebException e) {
+        Error error = new Error(e.getStatus().value(), e.getMessage());
+        return new ResponseEntity<>(error, e.getStatus());
+    }
+
+    @ExceptionHandler({IllegalArgumentException.class, IllegalStateException.class, MethodArgumentNotValidException.class})
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    private Error notValidArgumentHandler(Exception e) {
+        return new Error(HttpStatus.BAD_REQUEST.value(), "You need to initialize all fields");
+    }
+
+    private GenreDto convertToDto(Genre genre) {
+        return modelMapper.map(genre, GenreDto.class);
+    }
+
+    private Genre convertFromDto(GenreDto genreDto) {
+        return modelMapper.map(genreDto, Genre.class);
+    }
+
+    private String mapGenreName(String genreName) {
+        genreName = genreName.toLowerCase();
+        genreName = genreName.replaceAll("-", " ");
+        return genreName;
```

Comment:
What is the reason to do such kind of mapping? If it because of using genre name as a request param, then it can be eliminated.